ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
+
+		// Set the login style to SMF
+		if (empty($this->FSSettings['NameStyle']))
+			$this->FSSettings['NameStyle'] = 'smf';
+	}
+
+	/*
+	 * Read the Forum Settings file and load up settings for later on.
+	*/
+	public function readFSFSettings($file)
+	{
+		global $cookiename, $boardurl, $db_prefix, $db_type, $db_server, $db_user, $db_passwd, $db_name;
+
+		require ($this->FSSettings['path'] . '/Settings.php');
+
+		// Put these away.
+		foreach (array(
+			'cookiename',
+			'boardurl',
+			'db_prefix',
+			'db_type',
+			'db_server',
+			'db_user',
+			'db_passwd',
+			'db_name',
+			'auth_secret',
+			'cookie_no_auth_secret',
+			'image_proxy_secret',
+			'sourcedir'
+		) as $key)
+			$this->FSSettings[$key] = !empty($$key) ? $$key : null;
+	}
+
+	public function __construct(array $params = array())
+	{
+		global $wgSMFLogin, $wgFSPSoftware;
+
 		parent::__construct($params);
 
 		// TODO: Keep eye out for other things that need logging. Admin changes?
 		$this->logger = LoggerFactory::getInstance('ForumSessionProviderLog');
 
-        // Administrator always get admin rights.
-        if (!in_array(1, $GLOBALS['wgFSPAdminGroups']))
-            $GLOBALS['wgFSPAdminGroups'][] = 1;
+		// Set our software up.
+		$this->wgFSPSoftware = !empty($wgSMFLogin) ? 'smf2.0' : (!empty($wgFSPSoftware) ? $wgFSPSoftware : null);
+		$wgFSPSoftware = $this->wgFSPSoftware;
+
+		// Load our settings.
+		$this->loadFSSettings();
 
-        if (!in_array(1, $GLOBALS['wgFSPInterfaceGroups']))
-            $GLOBALS['wgFSPInterfaceGroups'][] = 1;
-		
-		if (is_readable($GLOBALS['wgFSPPath'] . '/Settings.php')) {
-		    require ($GLOBALS['wgFSPPath'] . '/Settings.php');
-		    // Globals declared in Settings.php gain local scope.
-		    $GLOBALS['wgFSPCookieName'] = $cookiename;
-		    $GLOBALS['wgFSPBoardURL'] = $boardurl;
+		// This is old Auth_SMF.php extension? Lets combat it.
+		if (!empty($wgSMFLogin) || in_array($wgFSPSoftware, array('smf2.0', 'smf2.1')))
+			$this->combatAuthSMF();
 
+		// Make sure we can find the settings file.
+		if (
+			!empty($this->FSSettings['path'])
+			&& is_readable($this->FSSettings['path'] . '/Settings.php')
+		)
+		{
+			// Read the Settings file in, use this layer to adjust what we need to bring in.
+			$this->readFSFSettings($this->FSSettings['path'] . '/Settings.php');
+
+			// Read the cookie
 			$this->decodeCookie();
 
-			if ($this->id && is_integer($this->id)) {
-                $this->prefix = $db_prefix;
+			// If we have a valid ID, lets connect to the database.
+			if ($this->FScookie['id'] && is_integer($this->FScookie['id']))
+				$this->FSDBConnect();
+		}
+		else
+			$this->logger->warning('Settings.php missing or not readable. Tried to load: ' . $this->FSSettings['path'] . '/Settings.php');
+
+		// Add in our AuthManager hack.
+		$this->AuthManagerHook();
+	}
+
+	/*
+	 * Inject the SMF AuthManager hook to override stuff we no longer need them to change
+	*/
+	private function AuthManagerHook()

```

Comment:
There is nothing in here unique to SMF (or any forum period), we should just hook it the proper way and block email/password changes, etc. raw.

Reply:
Sounds fair, let me do some testing on that.

Reply:
I did also do this so if we disable the extension, it becomes editable again.  Instead of having to disable it elsewhere as well.

Reply:
Woot. : )